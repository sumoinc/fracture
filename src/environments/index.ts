export * from "./environment";
export * from "./region";
