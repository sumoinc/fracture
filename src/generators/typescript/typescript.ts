import { SourceCode } from "projen";
import { FractureComponent } from "../../core/component";
import { Fracture } from "../../core/fracture";

export class TypeScript extends FractureComponent {
  constructor(fracture: Fracture) {
    super(fracture);
  }

  public preSynthesize() {
    /**
     * Build a TypeScript file and make sure it's marked as being managed by
     * projen.
     *
     * @param name
     * @returns {SourceCode}
     */
    const typeScriptFile = (name: string): SourceCode => {
      const f = new SourceCode(
        this.project,
        `${this.fracture.gendir}/types/${name}`
      );
      f.line("// " + f.marker);
      f.line("\n");
      return f;
    };

    // create index file
    const indexFile = typeScriptFile("index.ts");

    /**
     * Generate types for each entity.
     */
    this.fracture.entities.forEach((e) => {
      // add to index for export
      indexFile.line(`export * from "./${e.name}";`);

      // generate types for this entity
      const tsFile = typeScriptFile(`${e.name}.ts`);

      // generate helpful comments
      tsFile.line(`/**`);
      e.comment.forEach((c) => tsFile.line(` * ${c}`));
      tsFile.line(` */`);

      // build interface
      tsFile.open(`export interface ${e.name} {`);
      e.attributes.forEach((a) => {
        // generate helpful comments
        tsFile.line(`/**`);
        a.comment.forEach((c) => tsFile.line(` * ${c}`));
        tsFile.line(` */`);
        // console.log(a);

        // write type definition
        const q = a.isRequired ? "" : "?";
        tsFile.line(`${a.name}${q}: ${a.typeScriptType};`);
      });
      tsFile.close(`}`);
      tsFile.line("\n");
    });
    // close up index file
    indexFile.line("\n");
  }
}
