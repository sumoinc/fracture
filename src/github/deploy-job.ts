import { Component } from "projen";
import {
  Job,
  JobCallingReusableWorkflow,
  JobPermission,
} from "projen/lib/github/workflows-model";
import { Fracture } from "../core";

export interface DeployJobOptions {
  /**
   * Job ID for the deploy job.
   *
   * @default deploy
   */
  jobId: string;
  /**
   * The command to run to deploy the code.
   *
   * @default "npx projen deploy"
   */
  deployCommand: string;
}

export class DeployJob extends Component {
  /**
   * The job definition, suitable for adding to a workflow.
   */
  public readonly job: Job | JobCallingReusableWorkflow;
  /**
   * The Job Id for the deploy job
   *
   * @default deploy
   */
  public readonly jobId: string;
  /**
   * The command to run to deploy the code.
   *
   * @default "npx projen deploy"
   */
  public readonly deployCommand: string;

  constructor(fracture: Fracture, options: Partial<DeployJobOptions> = {}) {
    super(fracture);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.jobId = options.jobId ?? "deploy";
    this.deployCommand = options.deployCommand ?? "npx projen deploy";

    /***************************************************************************
     * Create job definition
     **************************************************************************/

    this.job = {
      runsOn: ["ubuntu-latest"],
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.WRITE,
      },
      steps: [
        {
          name: "Deploy",
          run: this.deployCommand,
        },
      ],
    };
  }
}
