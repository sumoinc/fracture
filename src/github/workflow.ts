import { paramCase } from "change-case";
import {
  GitHub,
  GithubWorkflow,
  GithubWorkflowOptions,
} from "projen/lib/github";
import { BuildJob } from "./build-job";
import { Fracture } from "../core";
import { TurboRepo } from "../turborepo";

export interface GithubFractureWorkflowOptions extends GithubWorkflowOptions {
  /**
   * The name of the workflow.
   */
  name: string;
  /**
   * What branches shoudl trigger this pipeline?
   */
  branchTriggerPatterns: string[];
  /**
   * What paths should trigger this pipeline?
   */
  pathTriggerPatterns: string[];
  /**
   * Pipelines always builds but don't always deploy.
   * Should this pipeline deploy the app after build?
   */
  deploy: boolean;
}

export class GithubFractureWorkflow extends GithubWorkflow {
  /**
   * The name of the workflow.
   */
  public readonly name: string;
  /**
   * Storage location for this workflow.
   */
  public readonly filePath: string;

  constructor(fracture: Fracture, options: GithubFractureWorkflowOptions) {
    /***************************************************************************
     * Super
     **************************************************************************/

    const github = GitHub.of(fracture);
    if (!github) {
      throw new Error(
        "Fracture workflows are currently only supported for GitHub enabled projects"
      );
    }
    const name = options.name ? paramCase(options.name) : "deployment";
    super(github, name, options);

    /***************************************************************************
     * Locals
     **************************************************************************/

    const turboRepo = TurboRepo.of(fracture);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.name = name;
    this.filePath = `.github/workflows/${this.name}.yml`;

    /***************************************************************************
     * Triggers
     **************************************************************************/

    this.on({
      push: {
        branches: options.branchTriggerPatterns,
        paths: options.pathTriggerPatterns,
      },
      workflowDispatch: {}, // allow manual triggering
    });

    /***************************************************************************
     *
     * BUILD JOB
     *
     * Leverage turbo-repo commands, if available.
     *
     **************************************************************************/

    const buildJob = !turboRepo
      ? new BuildJob(fracture)
      : new BuildJob(fracture, {
          buildCommand: fracture.runTaskCommand(turboRepo.buildTask),
          synthCommand: fracture.runTaskCommand(turboRepo.synthTask),
        });
    this.addJob(buildJob.jobId, buildJob.job);

    /***************************************************************************
     *
     * DEPLOY JOBS
     *
     **************************************************************************/

    // no waves, just use default workflow
    /*
    if (pipeline.waves.length === 0) {
      const deployJob = !turboRepo
        ? new DeployJob(fracture)
        : new DeployJob(fracture, {
            deployCommand: fracture.runTaskCommand(turboRepo.deployTask),
          });
      this.githubWorkflow.addJob(deployJob.jobId, deployJob.job);
    }*/

    // add deployment waves to workflow
    /*
    if (pipeline.waves.length > 0) {
      pipeline.waves.forEach((wave) => {
        wave.stages.forEach(() => {
          const deployJob = new DeployJob(pipeline.project as Fracture);
          this.githubWorkflow.addJob(deployJob.jobId, deployJob.job);
          //console.log(wave.name, stage.name);
          //this.addDeployJob(stage);
        });
      });
    }
    */
  }
}
