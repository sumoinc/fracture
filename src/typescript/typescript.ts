import { SourceCode } from "projen";
import { FractureComponent } from "../core/component";
import { Fracture } from "../core/fracture";

export class TypeScript extends FractureComponent {
  constructor(fracture: Fracture) {
    super(fracture);
  }

  public preSynthesize() {
    /**
     * Build a TypeScript file and make sure it's marked as being managed by
     * projen.
     *
     * @param name
     * @returns {SourceCode}
     */
    const typeScriptFile = (name: string): SourceCode => {
      const f = new SourceCode(
        this.project,
        `${this.fracture.gendir}/types/${name}`
      );
      f.line("// " + f.marker);
      f.line("\n");
      return f;
    };

    const indexFile = typeScriptFile("index.ts");
    this.fracture.entities.forEach((e) => {
      // add to index for export
      indexFile.line(`export * from "./${e.name}";`);

      // generate types for this entity
      const tsFile = typeScriptFile(`${e.name}.ts`);

      tsFile.open(`export interface ${e.name} {`);
      e.attributes.forEach((a) => {
        const q = a.isRequired ? "" : "?";
        tsFile.line(`${a.name}${q}: ${a.typeScriptType};`);
      });
      tsFile.close(`}`);
      tsFile.line("\n");
    });
    // close up index file
    indexFile.line("\n");
  }
}
