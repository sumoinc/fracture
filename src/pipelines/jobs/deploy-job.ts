import { paramCase } from "change-case";
import { Task } from "projen";
import { Job, JobPermission } from "projen/lib/github/workflows-model";
import { SetOptional, ValueOf } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { REGION_IDENTITIER } from "../../core";
import { renderDeploySteps } from "../steps/deploy-steps";
import { renderDownloadArtifactSteps } from "../steps/download-artifact-steps";
import { renderSetupNodeSteps } from "../steps/setup-node-steps";
import { Workflow } from "../workflow";

export interface DeployJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPattern?: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs?: Array<string>;

  /**
   * Projen Task that runs the deploy
   */
  readonly deployTask: Task;

  /**
   * AWS OIDC Credentials, if deploying to AWS
   */
  readonly awsOidcCredentials?: {
    roleToAssume: string;
    awsRegion: ValueOf<typeof REGION_IDENTITIER>;
    roleDurationSeconds: number;
  };
}

export class DeployJob extends WorkflowJob {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPattern: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs: Array<string>;

  /**
   * Projen Task that runs the deploy
   */
  readonly deployTask: Task;

  /**
   * AWS OIDC Credentials, if deploying to AWS
   */
  readonly awsOidcCredentials?: {
    roleToAssume: string;
    awsRegion: ValueOf<typeof REGION_IDENTITIER>;
    roleDurationSeconds: number;
  };

  constructor(public readonly workflow: Workflow, options: DeployJobOptions) {
    const branchPattern = options.branchPattern ?? "main";
    const name = options.name ?? `Deploy ${branchPattern}`;
    const jobId = options.jobId ?? paramCase(name);

    super(workflow, {
      jobId,
      name,
      ...options,
    });

    // defaults
    this.branchPattern = branchPattern;
    this.needs = [workflow.buildJob.jobId, ...(options.needs ?? [])];
    this.deployTask = options.deployTask;
    this.awsOidcCredentials = options.awsOidcCredentials;

    // make sure build stores these
    workflow.buildJob.artifactDirectories.push(...this.artifactDirectories);
  }

  dependsOn = (job: WorkflowJob): void => {
    this.needs.push(job.jobId);
  };

  render = (): Job => {
    return {
      name: this.name,
      runsOn: this.workflow.defaultRunners,
      needs: this.needs,
      if: `startsWith( github.ref, 'refs/heads/${this.branchPattern}' )`,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.READ,
        ...(this.awsOidcCredentials
          ? {
              idToken: JobPermission.WRITE,
            }
          : {}),
      },
      steps: [
        ...renderDownloadArtifactSteps(this),
        ...renderSetupNodeSteps(this.workflow),
        ...renderDeploySteps(this),
      ],
    };
  };
}
