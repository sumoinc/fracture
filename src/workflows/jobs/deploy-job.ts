import { paramCase } from "change-case";
import { Job, JobPermission, JobStep } from "projen/lib/github/workflows-model";
import { TypeScriptProject } from "projen/lib/typescript";
import { SetOptional } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { Environment } from "../../environments";
import { renderDeploySteps } from "../steps/deploy-steps";
import { renderDownloadArtifactSteps } from "../steps/download-artifact-steps";
import { renderSetupNodeSteps } from "../steps/setup-node-steps";
import { Workflow } from "../workflow";

export interface DeployJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {
  /**
   * How to identify the app being deployed
   *
   */
  readonly appName: string;

  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPrefix?: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs?: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  readonly artifactsDirectory: string;
}

export class DeployJob extends WorkflowJob {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  public readonly branchPrefix: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  public readonly needs: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  public readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  public readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  public readonly artifactsDirectory: string;

  /**
   * Workflow for deployment
   */
  readonly workflow: Workflow;

  constructor(
    public readonly project: TypeScriptProject,
    options: DeployJobOptions
  ) {
    const branchPrefix = options.branchPrefix ?? "main";
    const name = `Deploy ${options.appName} to ${options.environment.name}`;
    const jobId = options.jobId ?? paramCase(name);

    super(project, {
      jobId,
      name,
      ...options,
    });

    // defaults
    this.workflow = Workflow.deploy(project);
    this.branchPrefix = branchPrefix;
    this.needs = [this.workflow.buildJob.jobId, ...(options.needs ?? [])];
    this.deploySteps = options.deploySteps;
    this.environment = options.environment;
    this.artifactsDirectory = options.artifactsDirectory;

    // add this branch pattern as a trigger for the workflow
    this.workflow.addPushTrigger(this.branchPrefix);

    // make sure build step will store the artifacts we need
    this.workflow.buildJob.artifactDirectories.push(this.artifactsDirectory);
  }

  dependsOn = (job: WorkflowJob): void => {
    this.needs.push(job.jobId);
  };

  render = (): Job => {
    return {
      name: this.name,
      runsOn: this.workflow.defaultRunners,
      needs: this.needs,
      concurrency: this.jobId,
      if: `startsWith( github.ref, 'refs/heads/${this.branchPrefix}' )`,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.READ,
        ...(this.environment.authProviderType === "AWS_GITHUB_OIDC"
          ? {
              idToken: JobPermission.WRITE,
            }
          : {}),
      },
      steps: [
        ...renderDownloadArtifactSteps(this),
        ...renderSetupNodeSteps(this.workflow, {
          packageCache: false,
          setupPnpm: false,
        }),
        ...renderDeploySteps(this),
      ],
    };
  };
}
