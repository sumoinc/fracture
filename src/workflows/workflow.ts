import { Component } from "projen";
import { GitHub, GithubWorkflow } from "projen/lib/github";
import { TypeScriptProject } from "projen/lib/typescript";
import { BuildJob } from "./jobs/build-job";
import { DeployJob, DeployJobOptions } from "./jobs/deploy-job";
import { WorkflowJob } from "./jobs/workflow-job";

/**
 * Name of the permission back up file to include in the build artifact
 * to work around a GitHub Action bug that does not preserve file mode
 * permissions across upload and download actions.
 *
 * See {@link https://github.com/actions/upload-artifact/issues/38}
 */
export const PERMISSION_BACKUP_FILE = "permissions-backup.acl";

export interface WorkflowOptions {
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   *
   * @default workflow
   */
  readonly name?: string;

  /**
   * The node version to use in workflows.
   *
   * @default latest
   */
  readonly nodeVersion?: string;

  /**
   * Enable Node.js package cache in GitHub workflows.
   *
   * @default true
   */
  readonly packageCache?: boolean;

  /**
   * The version of PNPM to use if using PNPM as a package manager.
   *
   * @default "8"
   */
  readonly pnpmVersion?: string;

  /**
   * Automatically update files modified during builds to pull-request branches.
   * This means that any files synthesized by projen or e.g. test snapshots will
   * always be up-to-date before a PR is merged.
   *
   * Implies that PR builds do not have anti-tamper checks.
   *
   * This is enabled by default only if `githubTokenSecret` is set. Otherwise it
   * is disabled, which implies that file changes that happen during build will
   * not be pushed back to the branch.
   *
   * @default false
   */
  readonly mutableBuild?: boolean;

  /**
   * What runners should this workflow use by default?
   *
   * @default ["ubuntu-latest"]
   */
  readonly defaultRunners?: Array<string>;
}

export class Workflow extends Component {
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   */
  public readonly name: string;

  /**
   * The node version to use in workflows.
   *
   * @default latest
   */
  public readonly nodeVersion: string;

  /**
   * Enable Node.js package cache in GitHub workflows.
   *
   * @default true
   */
  public readonly packageCache: boolean;

  /**
   * Branch or tag name.
   * @default - the default branch is implicitly used
   */
  public readonly ref?: string;

  /**
   * The repository (owner/repo) to use.
   * @default - the default repository is implicitly used
   */
  public readonly repository?: string;

  /**
   * Whether to download files from Git LFS for this workflow
   *
   * @default - Use the setting on the corresponding GitHub project
   */
  public readonly downloadLfs?: boolean;

  /**
   * The version of PNPM to use if using PNPM as a package manager.
   *
   * @default "8"
   */
  public readonly pnpmVersion: string;

  /**
   * Automatically update files modified during builds to pull-request branches.
   * This means that any files synthesized by projen or e.g. test snapshots will
   * always be up-to-date before a PR is merged.
   *
   * Implies that PR builds do not have anti-tamper checks.
   *
   * This is enabled by default only if `githubTokenSecret` is set. Otherwise it
   * is disabled, which implies that file changes that happen during build will
   * not be pushed back to the branch.
   *
   * @default false
   */
  readonly mutableBuild: boolean;

  /**
   * What runners should this workflow use by default?
   *
   * @default ["ubuntu-latest"]
   */
  readonly defaultRunners: Array<string>;

  /**
   * Instance of github configuration for the project
   */
  public readonly github: GitHub;

  /**
   * Github workflow Workflow
   */
  public readonly workflow: GithubWorkflow;

  /**
   * Build Job for the workflow.
   */
  public readonly buildJob: BuildJob;

  /**
   * Other Jobs in this workflow
   */
  public readonly otherJobs: Array<WorkflowJob> = [];

  constructor(
    public readonly project: TypeScriptProject,
    options: WorkflowOptions
  ) {
    super(project);

    const github = GitHub.of(project);

    if (!github) {
      throw new Error(
        "Workflows are currently only supported for GitHub projects"
      );
    }

    // defaults
    this.name = options.name ?? "workflow";
    this.nodeVersion = options.nodeVersion ?? "latest";
    this.packageCache = options.packageCache ?? true;
    this.pnpmVersion = options.pnpmVersion ?? "8";
    this.mutableBuild = options.mutableBuild ?? false;
    this.defaultRunners = options.defaultRunners ?? ["ubuntu-latest"];

    // local
    this.github = github;
    this.workflow = new GithubWorkflow(github, this.name);

    // render the build steps
    this.buildJob = new BuildJob(this);
  }

  addDeployJob(options: DeployJobOptions): DeployJob {
    const job = new DeployJob(this, { ...options });
    this.otherJobs.push(job);
    return job;
  }

  preSynthesize(): void {
    // add the build job to the workflow
    this.workflow.addJob(this.buildJob.jobId, this.buildJob.render());
    // add other jobs
    this.otherJobs.forEach((job) => {
      this.workflow.addJob(job.jobId, job.render());
    });
  }
}
