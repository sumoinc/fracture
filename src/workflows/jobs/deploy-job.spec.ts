import { JobStep } from "projen/lib/github/workflows-model";
import { TypeScriptProject } from "projen/lib/typescript";
import { DeployJob } from "./deploy-job";
import { AwsEnvironment } from "../../environments";
import { Workflow } from "../workflow";

describe("success conditions", () => {
  test("Smoke test", () => {
    const project = new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const workflow = new Workflow(project, {
      name: "my-workflow",
    });
    const deploySteps: Array<JobStep> = [
      {
        name: "Say foo",
        run: "echo 'foo'",
      },
    ];
    const environment = new AwsEnvironment(project, {
      name: "my-environment",
      accountNumber: "0000000000",
    });
    const deployJob = new DeployJob(workflow, {
      deploySteps,
      environment,
    });
    expect(deployJob).toBeTruthy();
    expect(deployJob.render()).toMatchSnapshot();
    // console.log(deployJob.render());
  });

  test("Two deploys depend on each other", () => {
    const project = new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const workflow = new Workflow(project, {
      name: "my-workflow",
    });
    const deploySteps: Array<JobStep> = [
      {
        name: "Say foo",
        run: "echo 'foo'",
      },
    ];
    const environment = new AwsEnvironment(project, {
      name: "my-environment",
      accountNumber: "0000000000",
    });
    const deployJobOne = new DeployJob(workflow, {
      name: "Deploy Service One",
      deploySteps,
      environment,
    });
    const deployJobTwo = new DeployJob(workflow, {
      name: "Deploy Service Two",
      deploySteps,
      environment,
    });
    deployJobTwo.dependsOn(deployJobOne);
    expect(deployJobOne).toBeTruthy();
    expect(deployJobOne.render()).toMatchSnapshot();
    expect(deployJobTwo).toBeTruthy();
    expect(deployJobTwo.render()).toMatchSnapshot();
    // console.log(deployJobOne.jobId, deployJobOne.render());
    // console.log(deployJobTwo.jobId, deployJobTwo.render());
  });

  test("Deploy with artifacts", () => {
    const project = new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const workflow = new Workflow(project, {
      name: "my-workflow",
    });
    const deploySteps: Array<JobStep> = [
      {
        name: "Say foo",
        run: "echo 'foo'",
      },
    ];
    const environment = new AwsEnvironment(project, {
      name: "my-environment",
      accountNumber: "0000000000",
    });
    const deployJob = new DeployJob(workflow, {
      name: "Deploy Service Foo",
      artifactDirectories: ["foo", "bar"],
      deploySteps,
      environment,
    });
    expect(deployJob).toBeTruthy();
    expect(deployJob.render()).toMatchSnapshot();
    // console.log(JSON.stringify(deployJob.render(), null, 2));
  });
});
