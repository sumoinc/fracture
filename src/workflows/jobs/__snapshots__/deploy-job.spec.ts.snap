// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success conditions Deploy with artifacts 1`] = `
Object {
  "env": Object {
    "CI": "true",
  },
  "if": "startsWith( github.ref, 'refs/heads/main' )",
  "name": "Deploy Service Foo",
  "needs": Array [
    "build",
  ],
  "permissions": Object {
    "contents": "read",
    "idToken": "write",
  },
  "runsOn": Array [
    "ubuntu-latest",
  ],
  "steps": Array [
    Object {
      "name": "Download build artifacts for \\"foo\\"",
      "uses": "actions/download-artifact@v3",
      "with": Object {
        "name": "build-artifact-acbd18db",
        "path": "foo",
      },
    },
    Object {
      "continueOnError": true,
      "name": "Restore build artifact permissions for \\"foo\\"",
      "run": "setfacl --restore=permissions-backup.acl",
      "workingDirectory": "foo",
    },
    Object {
      "name": "Download build artifacts for \\"bar\\"",
      "uses": "actions/download-artifact@v3",
      "with": Object {
        "name": "build-artifact-37b51d19",
        "path": "bar",
      },
    },
    Object {
      "continueOnError": true,
      "name": "Restore build artifact permissions for \\"bar\\"",
      "run": "setfacl --restore=permissions-backup.acl",
      "workingDirectory": "bar",
    },
    Object {
      "name": "Setup Node.js",
      "uses": "actions/setup-node@v3",
      "with": Object {
        "node-version": "latest",
      },
    },
    Object {
      "name": "Configure AWS Credentials",
      "uses": "aws-actions/configure-aws-credentials@v2",
      "with": Object {
        "aws-region": "us-east-1",
        "role-duration-seconds": 900,
        "role-to-assume": "arn:aws:iam::0000000000:role/GitHubDeploymentOIDCRole",
      },
    },
    Object {
      "name": "Say foo",
      "run": "echo 'foo'",
    },
  ],
}
`;

exports[`success conditions Smoke test 1`] = `
Object {
  "env": Object {
    "CI": "true",
  },
  "if": "startsWith( github.ref, 'refs/heads/main' )",
  "name": "Deploy main",
  "needs": Array [
    "build",
  ],
  "permissions": Object {
    "contents": "read",
    "idToken": "write",
  },
  "runsOn": Array [
    "ubuntu-latest",
  ],
  "steps": Array [
    Object {
      "name": "Setup Node.js",
      "uses": "actions/setup-node@v3",
      "with": Object {
        "node-version": "latest",
      },
    },
    Object {
      "name": "Configure AWS Credentials",
      "uses": "aws-actions/configure-aws-credentials@v2",
      "with": Object {
        "aws-region": "us-east-1",
        "role-duration-seconds": 900,
        "role-to-assume": "arn:aws:iam::0000000000:role/GitHubDeploymentOIDCRole",
      },
    },
    Object {
      "name": "Say foo",
      "run": "echo 'foo'",
    },
  ],
}
`;

exports[`success conditions Two deploys depend on each other 1`] = `
Object {
  "env": Object {
    "CI": "true",
  },
  "if": "startsWith( github.ref, 'refs/heads/main' )",
  "name": "Deploy Service One",
  "needs": Array [
    "build",
  ],
  "permissions": Object {
    "contents": "read",
    "idToken": "write",
  },
  "runsOn": Array [
    "ubuntu-latest",
  ],
  "steps": Array [
    Object {
      "name": "Setup Node.js",
      "uses": "actions/setup-node@v3",
      "with": Object {
        "node-version": "latest",
      },
    },
    Object {
      "name": "Configure AWS Credentials",
      "uses": "aws-actions/configure-aws-credentials@v2",
      "with": Object {
        "aws-region": "us-east-1",
        "role-duration-seconds": 900,
        "role-to-assume": "arn:aws:iam::0000000000:role/GitHubDeploymentOIDCRole",
      },
    },
    Object {
      "name": "Say foo",
      "run": "echo 'foo'",
    },
  ],
}
`;

exports[`success conditions Two deploys depend on each other 2`] = `
Object {
  "env": Object {
    "CI": "true",
  },
  "if": "startsWith( github.ref, 'refs/heads/main' )",
  "name": "Deploy Service Two",
  "needs": Array [
    "build",
    "deploy-service-one",
  ],
  "permissions": Object {
    "contents": "read",
    "idToken": "write",
  },
  "runsOn": Array [
    "ubuntu-latest",
  ],
  "steps": Array [
    Object {
      "name": "Setup Node.js",
      "uses": "actions/setup-node@v3",
      "with": Object {
        "node-version": "latest",
      },
    },
    Object {
      "name": "Configure AWS Credentials",
      "uses": "aws-actions/configure-aws-credentials@v2",
      "with": Object {
        "aws-region": "us-east-1",
        "role-duration-seconds": 900,
        "role-to-assume": "arn:aws:iam::0000000000:role/GitHubDeploymentOIDCRole",
      },
    },
    Object {
      "name": "Say foo",
      "run": "echo 'foo'",
    },
  ],
}
`;
